client/src/chat/Chat.jsx

```diff
@@ -0,0 +1,180 @@
+import React, { useEffect, useRef, useState } from "react";
+import { usePlayer, useStageTimer, usePlayers } from "@empirica/core/player/classic/react";
+import { Loading } from "@empirica/core/player/react";
+import { MessageBubble } from "./MessageBubble";
+import { TextBar } from "./TextBar";
+import { reconstructChatState, getNextActionId } from "./chatUtils";
+
+export function Chat({
+  scope,
+  attribute = "chat",
+  showNickname = true,
+  showTitle = false,
+  reactionEmojisAvailable = [],
+  reactToSelf = true,
+  numReactionsPerMessage = 1,
+}) {
+  const player = usePlayer();
+  const players = usePlayers();
+  const progressLabel = player.get("progressLabel");
+  const stageTimer = useStageTimer();
+  const [messages, setMessages] = useState([]);
+  const scroller = useRef();
+
+  const currentPlayerPosition = player?.get("position");
+
+  // Reconstruct chat state from actions
+  useEffect(() => {
```

Comment:
new actions (from either player) don't seem to trigger a re-render, even though messages seem to be stored correctly. I can send messages, but then have to refresh the page to see them. I wonder if we need to extract the actions outside of the useEffect, and then reconstruct the state if the length changes?